### PR TITLE
Expose showButtonMenu of PopupMenuButtonState

### PR DESCRIPTION
**Pull request by creativecreatorormaybenot**
I created this PR a few weeks ago already (#42830), but accidentally destroyed that branch.

## Description

It can be useful to be able to call `showButtonMenu` of a `PopupMenuButton` programmatically instead of it only being called when the user taps the button. There already is a `showMenu` function, however, that would require obtaining the position of the button manually and passing the items again. By exposing `showButtonMenu`, it is very easy to show the menu at the right place with the right items manually.

## Tests

I added the following tests:

 * Ensuring that the exposed method can be called and works correctly.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by chunhtai** [comment]
May I know what is the use case of exposing the api? Is it trying to solve a specific issue that has to do it this way?

**Comment by creativecreatorormaybenot** [comment]
@chunhtai 

I am not sure where I used this, but sometimes you want to programmatically show a popup menu. This is also what the framework reflects today because [`showMenu`](https://api.flutter.dev/flutter/material/showMenu.html) exists on a global scope (as a function).  
Now, if I want to show the same menu that would be displayed when a user **would tap** a `PopupMenuButton`, i.e. with the same values, same locations, etc., it would be extremely cumbersome (and unnecessary) to rewrite everything that is already done in [`_PopupMenuButtonState.showButtonMenu`](https://github.com/flutter/flutter/blob/1464eb8f5feb151520d163293fef290a962a0ac7/packages/flutter/lib/src/material/popup_menu.dart#L1088).

An example would be feature discovery: if you want to show what options a `PopupMenuButton` has to a user, you need to be able to either force them to tap the button, which would be weird, or open it programmatically for them, which is what this PR allows for.

**Comment by chunhtai** [comment]
We generally avoid exposing private state unless we have to. This might be an overkill for feature preview if you only want to show user how the menu looks like. It will actually open the menu and triggers on select after closing it. For that use case, it might be better to print out a dummy one instead. However, I can see how there might be some other use cases this can be helpful.

The code looks fine. @Hixie do you have any suggestion on whether we should expose this?

**Comment by creativecreatorormaybenot** [comment]
@chunhtai 

> It will actually open the menu and triggers on select after closing it.

That was the idea - at least for what I meant - because I meant literally opening the button programmatically.  

It is just up to you (the team) to decide I suppose because there are no further arguments - I mean it is a very simple change (one character to be exact 🙂 ).  
I object to forced privacy as a concept in general (it would be way more helpful to just hide code from autocompletion and giving a warning instead of the compiler interfering for no reason), so I think that I am biased anyway.

Is there an easy way to have a custom version of the framework for a project? Currently, I need use `flutter upgrade --force` and manually adjust all the files that I want to use differently, e.g. exposing private values, and it is very cumbersome - this is also the reason why I open these PRs in the first place.

**Comment by creativecreatorormaybenot** [comment]
@chunhtai 

> It will actually open the menu and triggers on select after closing it.

That was the idea - at least for what I meant - because it should be literally open the button programmatically.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 1]
```diff
@@ -1034,10 +1034,57 @@ class PopupMenuButton<T> extends StatefulWidget {
...
+///   void showButtonMenu() => globalKey.currentState.showButtonMenu();
+/// }
+/// ```
+///
+/// {@end-tool}
+///
+/// Alternatively, you could also visit child elements of your parent widget to obtain
+/// the state if you know exactly that the widget is a direct descendant.
+class PopupMenuButtonState<T> extends State<PopupMenuButton<T>> {
+  /// This will show a popup menu with the buttons supplied to
```
Comment (chunhtai) [inline packages/flutter/lib/src/material/popup_menu.dart]: maybe change to 
```dart
/// A method to show a ...
```

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 2]
```diff
@@ -1034,10 +1034,57 @@ class PopupMenuButton<T> extends StatefulWidget {
...
+  /// [PopupMenuButton.itemBuilder] at the position of your [PopupMenuButton].
+  ///
+  /// This is called by the state itself when the user taps the button
+  /// and [PopupMenuButton.enabled] is set to `true`.
```
Comment (chunhtai) [inline packages/flutter/lib/src/material/popup_menu.dart]: we also mention that developer call this method to show the menu programatically.

**Review by chunhtai — COMMENTED** [review]
We may also need to think about what happens when developer call showButtonMenu while this widget is not enabled

**Comment by creativecreatorormaybenot** [comment]
@chunhtai Thank you for your review! I tried to implement your feedback.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 3]
```diff
@@ -1037,10 +1037,51 @@ class PopupMenuButton<T> extends StatefulWidget {
...
+///
+/// See [showButtonMenu] for a way to programmatically open the popup menu
+/// of your button state.
+class PopupMenuButtonState<T> extends State<PopupMenuButton<T>> {
+  /// A method to show a popup menu with the items supplied to
+  /// [PopupMenuButton.itemBuilder] at the position of your [PopupMenuButton].
+  ///
+  /// By default, it is called when the user taps the button and [PopupMenuButton.enabled]
+  /// is set to `true`.
+  /// Moreover, you can open the button by calling the method manually.
```
Comment (chunhtai) [inline packages/flutter/lib/src/material/popup_menu.dart]: would you remove the new line between line1052 and line1053?

File: packages/flutter/test/material/popup_menu_test.dart [hunk 4]
```diff
@@ -1197,6 +1197,43 @@ void main() {
     expect(rootObserver.menuCount, 1);
     expect(nestedObserver.menuCount, 0);
   });
+
+  testWidgets('PopupMenuButton calling showButtonMenu manually', (WidgetTester tester) async {
+    // The GlobalKey is required to call showButtonMenu on the button because
```
Comment (chunhtai) [inline packages/flutter/test/material/popup_menu_test.dart]: This comment seems redundant, would you remove this one?

**Review by chunhtai — APPROVED** [review]
LGTM, with minor nit.

**Comment by chunhtai** [comment]
It seems like the analyzer is complaining about the tag.
To resolve this, you will have to hard reset the branch to current master and reapply the change.

File: packages/flutter/lib/src/material/popup_menu.dart (hunk 3 above)
Comment (creativecreatorormaybenot) [reply packages/flutter/lib/src/material/popup_menu.dart]: There are two trailing spaces in line 1052, which will push the next line to a new line in Markdown. Do you think that it is not appropriate?

**Comment by creativecreatorormaybenot** [comment]
Moved to #50670.

**Pull request closed without merging**